modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/DifferentFamiliesFunctionArgumentsCoercionTest.java

```diff
@@ -0,0 +1,366 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.planner.datatypes;
+
+import static org.apache.ignite.internal.lang.IgniteStringFormatter.format;
+import static org.apache.ignite.internal.sql.engine.util.TypeUtils.native2relationalType;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.runtime.CalciteContextException;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.apache.ignite.internal.sql.engine.planner.datatypes.utils.DifferentFamiliesPair;
+import org.apache.ignite.internal.sql.engine.schema.IgniteSchema;
+import org.apache.ignite.internal.sql.engine.type.IgniteTypeFactory;
+import org.apache.ignite.internal.sql.engine.util.Commons;
+import org.apache.ignite.internal.sql.engine.util.SqlTestUtils;
+import org.apache.ignite.internal.testframework.IgniteTestUtils;
+import org.apache.ignite.internal.type.NativeType;
+import org.apache.ignite.internal.type.NativeTypeSpec;
+import org.apache.ignite.internal.type.NativeTypes;
+import org.junit.jupiter.api.Assumptions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * A set of test to verify behavior of type coercion for function arguments, when values belongs to the different type families.
+ */
+@SuppressWarnings({"ThrowableNotThrown", "DataFlowIssue"})
+public class DifferentFamiliesFunctionArgumentsCoercionTest extends BaseTypeCoercionTest {
+    private static final IgniteSchema SCHEMA = createSchema();
+
+    private static final List<String> FUNCTIONS_WITH_NUMERIC_PARAM = List.of(
+            "SUBSTRING('abc', {})", "POSITION('a' IN 'abc' FROM {})", "EXP({})", "POWER({}, 1)"
```

Comment:
We also have a `SUBSTR` function, and this test fails for it.

Reply:
another nice catch, thanks!

turns out calcite's `FamilyOperandTypeChecker` simply skips validation when function has optional parameters and invocation omits them. I revised codebase and found only three affected functions: `TRUNCATE`, `SUBSTR`, and `ROUND`. All these function has been added to `DifferentFamiliesFunctionArgumentsCoercionTest`.

Hope I haven't missed anything.